game/qsp.ksy

```diff
@@ -0,0 +1,353 @@
+meta:
+  id: qsp
+  title: Quest Soft Player (QSP) Game
+  application: Quest Soft Player
+  file-extension: qsp
+  endian: le
+  encoding: "utf-16le" # in fact "ucs-2le"
+  license: Unlicense
+doc: |
+  QSP is a player for text quests written in a QSP language.
+
+doc-ref:
+  - https://github.com/QSPFoundation/qsp/blob/master/help/gam_desc.txt
+
+seq:
+  - id: signature
+    contents: ['Q',0, 'S',0, 'P',0, 'G',0, 'A',0, 'M',0, 'E',0, 0x0D,0, 0x0A,0] #QSPGAME
+    if: is_new_format
+  - id: version
+    type: usual_str
+    if:  is_new_format
+  - id: password
+    type: rot5_str
+    if:  is_new_format
+  - id: locations_count_str
+    type:
+      switch-on: is_new_format
+      cases:
+        true: rot5_number_str
+        false: number_str
+  - id: locations
+    type: loc
+    repeat: expr
+    repeat-expr: locations_count
+instances:
+  is_new_format:
+    value: true
+  locations_count:
+    value: "(is_new_format?locations_count_str.as<rot5_number_str>.number:locations_count_str.as<number_str>.number)"
+types:
+  string_placeholder:
+    doc: |
+      Computes offset and size of the string terminated by \r\n
+      WARNING: some so(u)rcery follows
+    seq:
+      - id: placeholder
+        type: char
+        if: pos > -1 # to trigger getting and caching it
+      - size: 0
+        if: size > -1 # to trigger computing and caching it
+    types:
+      char:
+        seq:
+          - id: c
+            type: u2le
+            repeat: until
+            repeat-until: _ == 0x0D
+          - id: c2
+            type: u2le
+          - id: next
+            type: char
+            if: c2 != 0x0A
+    instances:
+      pos:
+        value: _io.pos # populated in the beginning
+      size_full:
+        value: _io.pos - pos
+      size:
+        value: size_full - 4 # final \r\n are unneeded
+    types:
+      char:
+        seq:
+          - id: c
+            type: u2le
+            repeat: until
+            repeat-until: _ == 0x0D
+          - id: c2
+            type: u2le
+          - id: next
+            type: char
+            if: c2 != 0x0A
+  usual_str:
+    seq:
+      - id: placeholder
+        type: string_placeholder
+
+    instances:
+      val:
+        io: _io
+        size: placeholder.size
+        pos: placeholder.pos
+        type: str
+  wrapped_str:
+    doc: just a str wrapped in a type to overcome a bug in KSC
+    seq:
+      - id: val
+        type: str
+        size-eos: true
+  rot5_str:
+    -orig-id: "$$кодированная строка" # 'encoded string', in Russian
+    doc: According to the official docks it allows to put multiple lines into a single line.
+    seq:
+      - id: placeholder
+        type: string_placeholder
+    instances:
+      val:
+        value: derot.val
+      derot:
+        size: placeholder.size
+        pos: placeholder.pos
+        type: wrapped_str
+        process: "rot(5, 2)"
```

Comment:
Ok, so this `rot(...)` is where things start getting ugly. Would you like/want to start another repo like [kaitai_compress](https://github.com/kaitai-io/kaitai_compress), but for general-purpose routines?

Reply:
Yes, wouod be nice. But keep in mind that for proper serialization we need interfaces.